src/components/nodes/boundaryEvent/boundaryEvent.vue

```diff
@@ -76,32 +81,87 @@ export default {
       this.shape.position(x - (width / 2), y - (width / 2));
       this.updateCrownPosition();
     },
-    attachBoundaryEventToTask() {
-      const task = this.getTaskUnderShape();
+    attachBoundaryEventToTask(task) {
+      const currentlyAttachedTask = this.shape.getParentCell();
 
-      if (!task) {
-        return;
+      if (currentlyAttachedTask) {
+        currentlyAttachedTask.unembed(this.shape);
       }
 
       task.embed(this.shape);
       this.node.definition.set('attachedToRef', task.component.node.definition);
       this.toggleInterruptingStyle(this.node.definition.cancelActivity);
       this.updateShapePosition(task);
+    },
+    moveBoundaryEventIfOverTask() {
+      const task = this.getTaskUnderShape();
 
-      this.shape.listenTo(this.paper, 'element:pointerup', cellView => {
+      if (!task) {
+        this.shape.position(this.validPosition.x, this.validPosition.y);
+        this.updateCrownPosition();
+        return;
+      }
+
+      if (task === this.shape.getParentCell()) {
+        this.updateShapePosition(task);
+      }
+
+      this.attachBoundaryEventToTask(task);
+    },
+    listenForValidDropTargets() {
+      this.shape.listenTo(this.paper, 'element:pointerdown', cellView => {
         if (cellView.model !== this.shape) {
           return;
         }
-        this.updateShapePosition(task);
+
+        this.validPosition = this.shape.position();
+        this.shape.listenToOnce(this.paper, 'cell:pointerup blank:pointerup', this.moveBoundaryEventIfOverTask);
+      });
+    },
+    isPoolShape(model) {
+      return model.component.node.type === 'processmaker-modeler-pool';
+    },
+    listenForInvalidDropTargets() {
+      this.shape.on('change:position', () => {
+        const targetElement = this.graph
+          .findModelsUnderElement(this.shape)
+          .filter(model => model.component)[0];
+
+        const targetHasNotChanged = this.invalidTargetElement === targetElement;
+        if (targetHasNotChanged) {
+          return;
+        }
+
+        const targetIsInvalid = targetElement && !this.isValidBoundaryEventTarget(targetElement);
+        if (targetIsInvalid) {
+          targetElement.attr('body/fill', invalidNodeColor);
+        }
+
+        if (this.invalidTargetElement) {
+          this.resetShapeColor(this.invalidTargetElement);
+        }
+
+        this.invalidTargetElement = targetElement;
       });
     },
+    resetShapeColor(shape) {
+      const defaultColor = this.isPoolShape(shape) ? poolColor : defaultNodeColor;
+      shape.attr('body/fill', defaultColor);
+    },
   },
   async mounted() {
     this.shape = new EventShape();
     this.setShapeProperties();
     this.shape.addTo(this.graph);
     await this.$nextTick();
-    this.attachBoundaryEventToTask();
+    this.listenForValidDropTargets();
+    this.listenForInvalidDropTargets();
```

Comment:
I wonder about the naming of these methods. They're about registering event handlers, could the naming reflect that? I also wonder if having this in multiple methods in this file also seems a bit smelly. Should these be in a separate file or at least consolidated somehow?

Reply:
> could the naming reflect that?

I was thinking that "listenFor" in the method name hints at this being related to registering an event listener. What naming convention would you suggest? "registerEventListeners"?

I'll look into consolidating the two functions, since they are very related, and I'll see if there's any logic I can pull out into a separate file.